### DISCUSSION TITLE
Mapping TRouter to remove an input param from all procedures

### DISCUSSION THREAD

**Discussion by dodas** [trpc/trpc#3993] [Q&A]
Hi!
I have a bit more specific ask / usecase:
I have a wrapper around how my routers are called from frontend and I esentially need to map the Router type to remove a certain property from all procedure inputs.

E.g. I have a router where all procedures accept `userId` input param (and possibly other params), but I want to exclude this prop from types, since my wrapper will add this prop on the fly, so it doesn't need to be passed in to individual queries.

```tsx
// original
const trpc = createTRPCReact<Router>();

trpc.files.list.useQuery({ userId: "abc" });

// I need to map Router so it doesn't include the `userId` params
const trpc = createTRPCReact<RemoveUserIdInput<Router>>();

trpc.files.list.useQuery();
```

Some experimenting, although not successful:

```tsx
type IsEmptyObject<Obj extends Record<PropertyKey, unknown>> = [
  keyof Obj
] extends [never]
  ? true
  : false;

type ProceduresMapped = {
  [Key in keyof Procedures]: Omit<Procedures[Key], "_def"> & {
    _def: Omit<Procedures[Key]["_def"], "_input_in"> & {
      _input_in: IsEmptyObject<
        Omit<Procedures[Key]["_def"]["_input_in"], "userId">
      > extends true
        ? typeof unsetMarker
        : Omit<Procedures[Key]["_def"]["_input_in"], "userId">;
    };
    (opts: ProcedureCallOptions): Promise<unknown>;
  };
};
```

Could anybody help with this?
Thanks!

**Comment by Nick-Lucas** [comment]
Why are you passing it in the first place? 

For auth needs you’d typically attach a token via headers and createContext or middleware then extracts the userId from the header and verifies it. tRPC supports this flow

**Reply by dodas** [reply]
This is really more like an input than auth, using `userId` was perhaps a bad example.
